File: 1034.边界着色.rs
/*
 * @lc app=leetcode.cn id=1034 lang=rust
 *
 * [1034] 边界着色
 */

// @lc code=start
use std::collections::VecDeque;

impl Solution {
    // bfs?
    pub fn color_border(mut grid: Vec<Vec<i32>>, row: i32, col: i32, color: i32) -> Vec<Vec<i32>> {
        let mut q = VecDeque::new();
        q.push_back((row as usize, col as usize));
        let mut visited = vec![vec![false; grid[0].len()]; grid.len()];
        visited[row as usize][col as usize] = true;
        let src_color = grid[row as usize][col as usize];
        let mut need_change = Vec::new();
        while let Some((r, c)) = q.pop_front() {
            // 首先我们检查这个点是否是边界, 如果是, 我们就将其染色
            if r == 0 || c == 0 || r == grid.len()-1 || c == grid[r].len() - 1 || grid[r-1][c] != src_color ||
            grid[r+1][c] != src_color || grid[r][c-1] != src_color || grid[r][c+1] != src_color {
                need_change.push((r, c));
            }
            // 然后我们考察后面四个方向
            if r - 1 < grid.len() && !visited[r-1][c] && grid[r-1][c] == src_color {
                visited[r-1][c] = true;
                q.push_back((r-1, c));
            }
            if r + 1 < grid.len() && !visited[r+1][c] && grid[r+1][c] == src_color {
                visited[r+1][c] = true;
                q.push_back((r+1, c));
            }
            if c - 1 < grid[r].len() && !visited[r][c-1] && grid[r][c-1] == src_color {
                visited[r][c-1] = true;
                q.push_back((r, c-1));
            }
            if c + 1 < grid[r].len() && !visited[r][c+1] && grid[r][c+1] == src_color {
                visited[r][c+1] = true;
                q.push_back((r, c+1));
            }
        }
        need_change.drain(..).for_each(|(r, c)| grid[r][c] = color);
        grid
    }
}
// @lc code=end

File: 1034.边界着色.js
/*
 * @lc app=leetcode.cn id=1034 lang=javascript
 *
 * [1034] 边界着色
 */

// @lc code=start
/**
 * @param {number[][]} grid
 * @param {number} row
 * @param {number} col
 * @param {number} color
 * @return {number[][]}
 */
 var colorBorder = function(grid, row, col, color) {
    let originColor = grid[row][col];
    if (originColor === color)  {
        return grid;
    }
    let cloneGride = JSON.parse(JSON.stringify(grid));
    let _y = grid.length-1;
    let _x = grid[0].length-1;
    let dfs = function(y, x) {
        if (grid[y][x] !== originColor) return;
        grid[y][x] *= -1;
        if (!(x !== 0 && x !== _x && y !== 0 && y !== _y && 
            Math.abs(grid[y-1][x]) === originColor && 
            Math.abs(grid[y+1][x]) === originColor && 
            Math.abs(grid[y][x-1]) === originColor && 
            Math.abs(grid[y][x+1]) === originColor) 
        ) {
        cloneGride[y][x] = color;
        }
        if (y > 0) dfs(y-1, x);
        if (x > 0) dfs(y, x-1);
        if (x < _x) dfs(y, x+1);
        if (y < _y) dfs(y+1, x);
    }
    dfs(row, col);
    return cloneGride;
};
// @lc code=end

// 常规的思路是可以使用深度优先搜索或者广度优先搜索来寻找出位置 (\textit{row},\textit{col})(row,col) 的所在的连通分量，额外要做的是搜索的时候需要判断当前的点是否属于边界。如果属于边界，需要把该点加入到一个用来存所有边界点的数组中。当搜索完毕后，再将所有边界点的进行着色。

